### DISCUSSION TITLE
Simple pyside6 program error on Mac M1

### DISCUSSION THREAD

**Discussion by paolodestefani** [pyinstaller/pyinstaller#7419] [Help] [answered]
Hi all
I'm new to macos builds and i'm teting how to use pyinstaller in this environment.
So i wrote small program a simple as this:

import sys

from PySide6 import __version__ as pyside6_version
from PySide6.QtWidgets import QWidget
from PySide6.QtWidgets import QApplication

print(f"You are using Python version {sys.version}")
print(f"PySide6 version is {pyside6_version}")

if __name__ == "__main__":
    "Start QApplication"
    app = QApplication(sys.argv)
    window = QWidget()
    window.show()
    sys.exit(app.exec())

Works perfectly launched from VSCodium or from a terminal.
I packed it with pyinstaller with not any error message but when i try to launch it i get the attached error message.
Can someone help me ?

I'm using MacOS 13.2 Ventura on a MacBook Air M1, python 3.11 and pyinstaller 5.7 installed using brew

[errorlog.txt](https://github.com/pyinstaller/pyinstaller/files/10668298/errorlog.txt)
[buildlog.txt](https://github.com/pyinstaller/pyinstaller/files/10668315/buildlog.txt)

**Comment by rokm** [comment, marked as answer]
> objc[34948]: Class QMacAutoReleasePoolTracker is implemented in both /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09/QtCore (0x10a15c5d8) and /opt/homebrew/Cellar/qt/6.4.2_1/lib/QtCore.framework/Versions/A/QtCore (0x1100085d8). One of the two will be used.

Looks like system copies of Qt libraries are leaking into the frozen application, and end up being used in addition to the copies collected into the frozen application, which leads to the crash. This happens either because we failed to overwrite a path on one of collected shared libraries *or* because we failed to collect some shared library, and therefore system copy is used (which pulls in other system copies). I suspect it's the second case.

FWIW, we test PySide6 PyPI wheels only, so homebrew-installed PySide6+Qt6 support might be lacking.

**Reply by rokm** [reply]
Aha, the plugins directory is not set up correctly (probably because homebrew PySide6 does not include embedded `qt.conf` to override plugins directory), so it ends up loading system-installed plugins, which pull in system-installed Qt libs.

Does overriding plugins directory like this work for you?

```python
import sys

if getattr(sys, 'frozen', False):
    import os
    plugins_dir = os.path.join(sys._MEIPASS, 'PySide6', 'Qt', 'plugins')
    print(f"Redirecting plugins dir: {plugins_dir}")
    os.environ['QT_PLUGIN_PATH'] = plugins_dir

from PySide6 import __version__ as pyside6_version
from PySide6.QtWidgets import QWidget
from PySide6.QtWidgets import QApplication

print(f"You are using Python version {sys.version}")
print(f"PySide6 version is {pyside6_version}")

if __name__ == "__main__":
    app = QApplication(sys.argv)
    window = QWidget()
    window.show()
    sys.exit(app.exec())
```

**Reply by paolodestefani** [reply]
Thanks for the quick reply.
I suppose the same BUT i didn't find any Qt libraries outside /opt/homebrew/(...)

paolo@macbook HelloWorld % cd /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09
cd: no such file or directory: /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09
paolo@macbook HelloWorld % cd /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T
paolo@macbook T % pwd
/private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T
paolo@macbook T % ls -la _*
zsh: no matches found: _*
paolo@macbook T % ls -la
total 88
drwx------@ 260 paolo  staff   8320  6 Feb 22:26 .
drwxr-xr-x@   7 paolo  staff    224  6 Feb 22:25 ..
drwx------    4 paolo  staff    128  6 Feb 03:44 .AddressBookLocks
-rw-r--r--@   1 paolo  staff   6148  6 Feb 22:29 .DS_Store
drwxr-xr-x    5 paolo  staff    160  6 Feb 22:18 .LINKS
drwx------@   4 paolo  staff    128  6 Feb 19:39 .com.google.Chrome.1pYU8a
-rw-r--r--@   1 paolo  staff      0  6 Feb 05:19 .keystoneAgentLock
drwx------    2 paolo  staff     64 24 Gen 20:29 AudioComponentRegistrar
-rw-------    1 paolo  staff      4  6 Feb 19:34 SOSBackup-BackupBagV0-tomb
-rw-------    1 paolo  staff      4  6 Feb 19:34 SOSBackup-iCloudIdentity-tomb
drwx------    2 paolo  staff     64 24 Gen 20:30 SandboxHelper
drwx------    3 paolo  staff     96 24 Gen 20:30 StatusKitAgent
drwx------@   2 paolo  staff     64  6 Feb 22:30 TemporaryItems
drwxr-xr-x@   2 paolo  staff     64 24 Gen 20:30 app_fs_ephemeral_files
drwxr-xr-x@   2 paolo  staff     64 24 Gen 20:30 app_fs_new_files
drwx------@   2 paolo  staff     64 24 Gen 21:01 com.apple.AMPArtworkAgent
drwx------@   2 paolo  staff     64 24 Gen 20:30 com.apple.AMPDeviceDiscoveryAgent
drwx------@   2 paolo  staff     64 24 Gen 20:30 com.apple.AirPlayUIAgent


i use homebrew because i read it's the recomanded way to install software on macos

**Reply by bwoodsend** [reply]
> i use homebrew because i read it's the recomanded way to install software on macos

Its Python support is terrible though.

**Reply by rokm** [reply]
> Thanks for the quick reply. I suppose the same BUT i didn't find any Qt libraries outside /opt/homebrew/(...)
> 
> paolo@macbook HelloWorld % cd /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09 cd: no such file or directory: /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09 paolo@macbook HelloWorld 

That's because you are using `onefile` build, so files are extracted (on each run) to unique temporary directory and removed once the program finishes (or crashes). So you won't find those files except while program is running.

**Reply by paolodestefani** [reply]
With the modified code works:

paolo@macbook HelloWorld % cd dist
paolo@macbook dist % ls -l
total 95424
-rwxr-xr-x@ 1 paolo  staff  48856176  6 Feb 22:48 HelloWorld
drwxr-xr-x@ 3 paolo  staff        96  6 Feb 22:48 HelloWorld.app
paolo@macbook dist % ./HelloWorld
Redirecting plugins dir: /var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIAan5QZ/PySide6/Qt/plugins
You are using Python version 3.11.1 (main, Dec 23 2022, 09:28:24) [Clang 14.0.0 (clang-1400.0.29.202)]
PySide6 version is 6.4.2
paolo@macbook dist %


So what do you suggest ? Remove everything python related and install from python.org/pypi.org/pip ?

**Reply by paolodestefani** [reply]
> > Thanks for the quick reply. I suppose the same BUT i didn't find any Qt libraries outside /opt/homebrew/(...)
> > paolo@macbook HelloWorld % cd /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09 cd: no such file or directory: /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09 paolo@macbook HelloWorld
> 
> That's because you are using `onefile` build, so files are extracted (on each run) to unique temporary directory and removed once the program finishes (or crashes). So you won't find those files except while program is running.

If i use onedir i get a similar error (without your hint):

paolo@macbook HelloWorld % ./HelloWorld
You are using Python version 3.11.1 (main, Dec 23 2022, 09:28:24) [Clang 14.0.0 (clang-1400.0.29.202)]
PySide6 version is 6.4.2
objc[37224]: Class QMacAutoReleasePoolTracker is implemented in both /Users/paolo/Development/Dist/HelloWorld/dist/HelloWorld/QtCore (0x1062a85d8) and /opt/homebrew/Cellar/qt/6.4.2_1/lib/QtCore.framework/Versions/A/QtCore (0x10c1545d8). One of the two will be used. Which one is undefined.

**Reply by rokm** [reply]
> > > Thanks for the quick reply. I suppose the same BUT i didn't find any Qt libraries outside /opt/homebrew/(...)
> > > paolo@macbook HelloWorld % cd /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09 cd: no such file or directory: /private/var/folders/1l/jn3pw75n0wsg2m9srlqdjh8r0000gq/T/_MEIHkSn09 paolo@macbook HelloWorld
> > 
> > 
> > That's because you are using `onefile` build, so files are extracted (on each run) to unique temporary directory and removed once the program finishes (or crashes). So you won't find those files except while program is running.
> 
> If i use onedir i get a similar error (without your hint):

Yes, because regardless of the build type, plugins directory is not set up correctly at runtime (without the path being explicitly overridden via environment variable). I was referring to why you could not find those libraries in the temporary directory (and temporary directory itself as well).

**Reply by rokm** [reply]
> So what do you suggest ? Remove everything python related and install from python.org/pypi.org/pip ?

It's up to you. If you want to stick with homebrew, you'll need that `QT_PLUGIN_PATH` override at the start of your program. 

There's no guarantees you won't run into additional (and different) issues with one or the other, though...

**Reply by paolodestefani** [reply]
So i tried the new way: remove pyinstaller/pyside/python from homebrew and install python 3.11 from python.org/download, create a virtualenv, install pyside6 and pyinstaller using pip

I can confirm that everything is working as aspected now.

Thanks
